frontend/src/reducers/cartReducers.js

```diff
@@ -0,0 +1,33 @@
+import {CART_ADD_ITEM, CART_REMOVE_ITEM} from '../constants/cartConstants'
+
+export const cartReducer = (state = {cartItems:[]},action) =>{
+    switch(action.type){
+        case CART_ADD_ITEM:
+            const item = action.payload
+            const existItem = state.cartItems.find((x) => x.product === item.product)
+            
+            if(existItem){
+                return {
+                    ...state,
+                    cartItems:state.cartItems.map((x) =>
+                    x.product === existItem.product ? item:x
+
+                    )
+                }
+            }else{
+                return{
+                    ...state,
+                    cartItems: [...state.cartItems,item]
+
+                }
+            }
```

Comment:
if i were you i would do this condition outside the reducer, 
like a helper function returns the output directly to keep the reducer clean

Reply:
Do you mean sagas methods ?

Reply:
no i mean like a helper method 
so instead of 

```js
        case CART_ADD_ITEM:
            const item = action.payload
            const existItem = state.cartItems.find((x) => x.product === item.product)
           if(existItem){
                return {
                    ...state,
                    cartItems:state.cartItems.map((x) =>
                    x.product === existItem.product ? item:x

                    )
                }
            }else{
                return{
                    ...state,
                    cartItems: [...state.cartItems,item]

              }
        }
            
```

we may have something like 

```js

case actions.ADD_ITEM:
            return { ...state, items: setCartItems(state.items, action.payload.data) }


// outside the reducer
function setCartItems(allItems = [], newItem) {
    let itemExists = allItems.find(item => item._id === newItem._id)
    let items = allItems

    if (itemExists) {
        items = items.map((item) => {
            if (item._id === itemExists._id) {
                item.qyt = item.qyt + 1
            }
            return item
        })
    } else {
        items = [...items, { ...newItem, qyt: 1 }]
    }
    localStorage.setItem("cartItems", JSON.stringify(items))
    return items
}
```

my only objective is to keep the reducer neat

Reply:
Okay I'm gonna edit it and thank you